File: frameworks/bridge/declarative_frontend/ark_modifier/src/marquee_modifier.ts
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class LazyArkMarqueeComponent extends ArkComponent {
  static module: MarqueeComponentModule | undefined = undefined;
  	constructor(nativePtr: KNode, classType: ModifierType) {
  	  super(nativePtr, classType);
  	  if (LazyArkMarqueeComponent.module === undefined) {
  	    LazyArkMarqueeComponent.module = globalThis.requireNapi('arkui.components.arkmarquee');
  	  }
  	  this.lazyComponent = LazyArkMarqueeComponent.module.createComponent(nativePtr, classType);
  	}
  setMap(): void {
   this.lazyComponent._modifiersWithKeys = this._modifiersWithKeys;
  }

  fontSize(value: Length): LazyArkMarqueeComponent {
   this.lazyComponent.fontSize(value);
   return this;
  }

  fontColor(value: ResourceColor): LazyArkMarqueeComponent {
   this.lazyComponent.fontColor(value);
   return this;
  }

  allowScale(value: boolean): LazyArkMarqueeComponent {
   this.lazyComponent.allowScale(value);
   return this;
  }

  fontWeight(value: number | FontWeight | string): LazyArkMarqueeComponent {
   this.lazyComponent.fontWeight(value);
   return this;
  }

  fontFamily(value: string | Resource): LazyArkMarqueeComponent {
   this.lazyComponent.fontFamily(value);
   return this;
  }

  onStart(event: () => void): LazyArkMarqueeComponent {
   this.lazyComponent.onStart(event);
   return this;
  }

  onBounce(event: () => void): LazyArkMarqueeComponent {
   this.lazyComponent.onBounce(event);
   return this;
  }

  onFinish(event: () => void): LazyArkMarqueeComponent {
   this.lazyComponent.onFinish(event);
   return this;
  }

  marqueeUpdateStrategy(value: MarqueeUpdateStrategy): LazyArkMarqueeComponent {
   this.lazyComponent.marqueeUpdateStrategy(value);
   return this;
  }
}
class MarqueeModifier extends LazyArkMarqueeComponent implements AttributeModifier<MarqueeAttribute> {

  constructor(nativePtr: KNode, classType: ModifierType) {
    super(nativePtr, classType);
    this._modifiersWithKeys = new ModifierMap();
    this.setMap();
  }

  applyNormalAttribute(instance: MarqueeAttribute): void {
    ModifierUtils.applySetOnChange(this);
    ModifierUtils.applyAndMergeModifier<MarqueeAttribute, ArkMarqueeComponent, ArkComponent>(instance, this);
  }
}

File: frameworks/core/interfaces/native/implementation/marquee_modifier.cpp
/*
 * Copyright (c) 2024-2025 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "base/log/log_wrapper.h"
#include "core/interfaces/native/generated/interface/arkoala_api_generated.h"
#include "core/common/dynamic_module_helper.h"
#include "ui/base/utils/utils.h"

namespace OHOS::Ace::NG::GeneratedModifier {
#ifdef ARKUI_CAPI_UNITTEST
const GENERATED_ArkUIMarqueeModifier* GetMarqueeStaticModifier();
#endif
const GENERATED_ArkUIMarqueeModifier* GetMarqueeModifier()
{
    static const GENERATED_ArkUIMarqueeModifier* cachedModifier = nullptr;

    if (cachedModifier == nullptr) {
#ifdef ARKUI_CAPI_UNITTEST
        cachedModifier = GeneratedModifier::GetMarqueeStaticModifier();
#else
        auto* module = DynamicModuleHelper::GetInstance().GetDynamicModule("Marquee");
        CHECK_NULL_RETURN(module, nullptr);
        cachedModifier = reinterpret_cast<const GENERATED_ArkUIMarqueeModifier*>(module->GetStaticModifier());
#endif
    }

    return cachedModifier;
}

}
